<a href="https://colab.research.google.com/github/skj092/experiments/blob/main/Activation_function_and_Weight_initializer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Objective : 

**To Study different activation functions and weight initializers**

In [1]:
# !pip install mlflow -q

In [2]:
# TensorFlow and tf.keras
import tensorflow as tf

# Helper libraries
import numpy as np
from tensorflow.keras import initializers
from tensorflow.python.keras import activations

print(tf.__version__)

# downloading fashion_mnist data
fashion_mnist = tf.keras.datasets.fashion_mnist

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
               'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

train_images = train_images / 255.0

test_images = test_images / 255.0              


import mlflow.tensorflow
mlflow.tensorflow.autolog()

c:\users\sonu.ramkumar.jha\desktop\experiments\env\lib\site-packages\numpy\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
c:\users\sonu.ramkumar.jha\desktop\experiments\env\lib\site-packages\numpy\.libs\libopenblas.GK7GX5KEQ4F6UYO3P26ULGBQYHGQO7J4.gfortran-win_amd64.dll
c:\users\sonu.ramkumar.jha\desktop\experiments\env\lib\site-packages\numpy\.libs\libopenblas.WCDJNK7YVMPZQ2ME2ZZHJJRJ3JIKNDB7.gfortran-win_amd64.dll
  warnings.warn("loaded more than 1 DLL from .libs:"


2.5.0


2021/06/14 14:33:16 WARNING mlflow.utils.autologging_utils: You are using an unsupported version of tensorflow. If you encounter errors during autologging, try upgrading / downgrading tensorflow to a supported version, or try upgrading MLflow.


# With Default Weight Initializer

## 1. relu

In [3]:
activation = tf.keras.activations.relu

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation),
tf.keras.layers.Dense(10)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)
    mlflow.tensorflow.mlflow.log_param('activation', activation)

del model

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten (Flatten)            (None, 784)               0         
_________________________________________________________________
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dense_1 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.4906 - accuracy: 0.8268
Epoch 2/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.3733 - accuracy: 0.8659
Epoch 3/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.3345 - accuracy: 0.8781
Epoch 4/10
1875/1875 [==============================] - 3s 1ms/step - loss

## 2. Softmax

In [4]:
activation = tf.keras.activations.softmax

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation),
tf.keras.layers.Dense(10)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('initializer', model.layers[1].kernel_initializer)
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    
del model

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_1 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_3 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 4s 2ms/step - loss: 1.3811 - accuracy: 0.6191
Epoch 2/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.8298 - accuracy: 0.6755
Epoch 3/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.7209 - accuracy: 0.6961
Epoch 4/10
1875/1875 [==============================] - 3s 1ms/step - lo

## 3. Sigmoid

In [5]:
activation = tf.keras.activations.sigmoid

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation),
tf.keras.layers.Dense(10)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('initializer', model.layers[1].kernel_initializer)
    mlflow.tensorflow.mlflow.log_param('activation', activation)

del model

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_2 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_4 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_5 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.5405 - accuracy: 0.8150
Epoch 2/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3920 - accuracy: 0.8594
Epoch 3/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3544 - accuracy: 0.8731
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - lo

## 4. tanh

In [6]:
activation = tf.keras.activations.tanh

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation),
tf.keras.layers.Dense(10)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('initializer', model.layers[1].kernel_initializer)
    mlflow.tensorflow.mlflow.log_param('activation', activation)

del model

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_3 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_6 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_7 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.4769 - accuracy: 0.8298
Epoch 2/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.3697 - accuracy: 0.8636
Epoch 3/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3352 - accuracy: 0.8766
Epoch 4/10
1875/1875 [==============================] - 3s 1ms/step - lo

## 5. selu

In [7]:
activation = tf.keras.activations.selu

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation),
tf.keras.layers.Dense(10)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('initializer', model.layers[1].kernel_initializer)
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    
del model

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_4 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_8 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_9 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.4923 - accuracy: 0.8239
Epoch 2/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.3857 - accuracy: 0.8605
Epoch 3/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3516 - accuracy: 0.8716
Epoch 4/10
1875/1875 [==============================] - 4s 2ms/step - lo

# With Weight Initializer

## 1. relu + he_normal

In [8]:
activation = tf.keras.activations.relu
initializer = tf.keras.initializers.he_normal

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation, kernel_initializer=initializer),
tf.keras.layers.Dense(10,kernel_initializer=initializer)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)

del model

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_5 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_10 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_11 (Dense)             (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 2ms/step - loss: 0.4947 - accuracy: 0.8256
Epoch 2/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.3712 - accuracy: 0.8656
Epoch 3/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3310 - accuracy: 0.8791
Epoch 4/10
1875/1875 [==============================] - 3s 2ms/step - lo

## 2. softmax + glorot_normal

In [9]:
activation = tf.keras.activations.softmax
initializer = tf.keras.initializers.glorot_normal

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation, kernel_initializer=initializer),
tf.keras.layers.Dense(10,kernel_initializer=initializer)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)
    
del model

Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_6 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_12 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_13 (Dense)             (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 5s 3ms/step - loss: 1.3332 - accuracy: 0.6831
Epoch 2/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.7282 - accuracy: 0.7452
Epoch 3/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.6123 - accuracy: 0.7541
Epoch 4/10
1875/1875 [==============================] - 3s 2ms/step - lo

## 3. sigmoid + glorot_mormal

In [10]:
initializer = tf.keras.initializers.glorot_normal
activation = tf.keras.activations.sigmoid

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation, kernel_initializer=initializer),
tf.keras.layers.Dense(10,kernel_initializer=initializer)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)
    
del model

Model: "sequential_7"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_7 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_14 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_15 (Dense)             (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.5434 - accuracy: 0.8130
Epoch 2/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3899 - accuracy: 0.8587
Epoch 3/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3529 - accuracy: 0.8721
Epoch 4/10
1875/1875 [==============================] - 2s 1ms/step - lo

## 4. tanh + glorot_normal

In [11]:
initializer = tf.keras.initializers.glorot_normal
activation = tf.keras.activations.tanh

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation, kernel_initializer=initializer),
tf.keras.layers.Dense(10,kernel_initializer=initializer)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)
    
del model

Model: "sequential_8"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_8 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_16 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_17 (Dense)             (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.4740 - accuracy: 0.8293
Epoch 2/10
1875/1875 [==============================] - 3s 1ms/step - loss: 0.3700 - accuracy: 0.8655
Epoch 3/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3343 - accuracy: 0.8774
Epoch 4/10
1875/1875 [==============================] - 3s 2ms/step - lo

## 5. selu + lucan_normal

In [12]:
activation = tf.keras.activations.selu
initializer = tf.keras.initializers.lecun_normal

initializer = tf.keras.initializers.glorot_normal
activation = tf.keras.activations.tanh

model = tf.keras.Sequential([
tf.keras.layers.Flatten(input_shape=(28, 28)),
tf.keras.layers.Dense(128, activation=activation, kernel_initializer=initializer),
tf.keras.layers.Dense(10,kernel_initializer=initializer)
])

model.compile(optimizer='adam',loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),metrics=['accuracy'])

# model summary
model.summary()

with mlflow.start_run():

    model.fit(train_images, train_labels, epochs=10)

    test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

    print('test_loss', test_loss)
    print('test_accuracy', test_acc)

    mlflow.tensorflow.mlflow.log_metric('test_loss', test_loss)
    mlflow.tensorflow.mlflow.log_metric('test_acc', test_acc)  
    mlflow.tensorflow.mlflow.log_param('activation', activation)
    mlflow.tensorflow.mlflow.log_param('initializer',model.layers[1].kernel_initializer)
    
del model

Model: "sequential_9"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_9 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_18 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_19 (Dense)             (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________
Epoch 1/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.4778 - accuracy: 0.8292
Epoch 2/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3683 - accuracy: 0.8666
Epoch 3/10
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3371 - accuracy: 0.8763
Epoch 4/10
1875/1875 [==============================] - 2s 1ms/step - lo

# Results : 

|Act Func|initializer|train_loss|train_accuracy|test_loss|test_accuracy|
|----|----|----|----|----|----|
|relu|GlorotUniform|0.2378 |0.9109	|0.3364 |0.8807|
|relu|he_normal|0.2423 |0.9089	|0.3358 |0.8867|
|softmax|GlorotUniform|0.4847 |0.8383	|0.5149 |0.8275|
|softmax|glorot_normal|0.4730 |0.8419	|0.5077 |0.8318|
|sigmoid|GlorotUniform|0.2503 |0.9078	|0.3451 |0.8767|
|sigmoid|glorot_normal|0.2517 |0.9063	|0.3325 |0.8791|
|tanh|GlorotUniform|0.2396 |0.9101	|0.3354 |0.8812|
|tanh|glorot_normal|0.2433 |0.9086	|0.3297 |0.8811|
|selu|GlorotUniform|0.2470 |0.9064	|0.3470 |0.8791|
|selu|lucan_normal|0.2416 |0.9112	|0.3346 |0.8812|

**Observations**
- By Default Dense layer have `Glorot Uniform` weight initializer.
- `Softmax` performed `worst` with both the initializers.
- Default weight initializer (GlorotUniform) performed worse in every activation function.
- Order of model according to test_loss is : `tanh > selu > sigmoid > relu > softmax`.
- Order of model according to test_accuracy is : `relu > tanh > selu > sigmoid > softmax`.